src/core/sdl.rs

```diff
@@ -363,48 +339,97 @@ fn print_directive_type_def(directive: &DirectiveDefinition) -> String {
     }
 }
 
-pub fn print(sd: ServiceDocument) -> String {
-    // Separate the definitions by type
-    let definitions_len = sd.definitions.len();
-    let mut schemas = Vec::with_capacity(definitions_len);
-    let mut scalars = Vec::with_capacity(definitions_len);
-    let mut interfaces = Vec::with_capacity(definitions_len);
-    let mut objects = Vec::with_capacity(definitions_len);
-    let mut enums = Vec::with_capacity(definitions_len);
-    let mut unions = Vec::with_capacity(definitions_len);
-    let mut inputs = Vec::with_capacity(definitions_len);
-    let mut directives = Vec::with_capacity(definitions_len);
-
-    for def in sd.definitions.iter() {
-        match def {
-            TypeSystemDefinition::Schema(schema) => schemas.push(print_schema(&schema.node)),
-            TypeSystemDefinition::Type(type_def) => match &type_def.node.kind {
-                TypeKind::Scalar => scalars.push(print_type_def(&type_def.node)),
-                TypeKind::Interface(_) => interfaces.push(print_type_def(&type_def.node)),
-                TypeKind::Enum(_) => enums.push(print_type_def(&type_def.node)),
-                TypeKind::Object(_) => objects.push(print_type_def(&type_def.node)),
-                TypeKind::Union(_) => unions.push(print_type_def(&type_def.node)),
-                TypeKind::InputObject(_) => inputs.push(print_type_def(&type_def.node)),
-            },
-            TypeSystemDefinition::Directive(type_def) => {
-                directives.push(print_directive_type_def(&type_def.node))
-            }
+#[derive(Default)]
+pub struct SdlPrinter {
+    pub federation_compatibility: bool,
+}
```

Comment:
This is an interesting approach where we by default add federation specific items if the flag is set but then we are scattering this sub-graph specific logic everywhere in our codebase. My recommendation is to add a top-level `definitions` field in `Config` where we can add any additional definitions, like in this case - `@link` these additional definitions aren't consumed by Tailcall but are outputed as it is.

Reply:
not exactly, the approach is not to output the Tailcall directives that conflicts with federation directives. To make the sdl output federation compatible we need to include apollo spec with federation's `@link` and by doing so we can't now leave tailcall's usage and definitions of `@link`